### PR TITLE
Add route to load full mesh stl of a segment/agglomerate

### LINKED ISSUE #7529: Unified Mesh Loading API
There are currently three different APIs for meshes: precomputed v3, precomputed (legacy), ad-hoc.

Can we provide a unified API for all three?

Things to consider
 - LODs
 - mags
 - multiple meshfiles can exist for the same segmentation
 - encodings/compression (DRACO,stl,…)
 - mappings (some meshfiles are computed for oversegmentation. need to request all the meshes mapped to target id)

Rough Design:
- LIST SOURCES: List available mesh sources (meshfiles, adhoc, remote?)
- METADATA + PARAMETER RANGES: Second route (or as part of first one) to query metadata for a mesh source:
  - available parameter ranges 
    - Numeric (LODs, mags, subsamplingStrides)
    - Applyable Mappings (mapping name in case of mapping-specific meshfile, ANY in case of oversegmentation meshfile or adhoc mapping)
  - encoding info (DRACO,stl,…)
  - needs NEIGHBORS logic
- LIST CHUNKS
  - For a mesh source, and a segment/agglomerate id, a mag, an LOD, a mapping name, list available chunks (returns transform, offset and for each chunk a position)
- CHUNK DATA
  - data for a list of chunks (as identified by mesh source, segmentId/agglomerateId, mappingName, lod, mag, position)
  - may return more chunks than are requested (in case of mapping and oversegmentation meshfile)
  - needs to return metadata as well (byte lengths of returned chunks + counts per position)

Problem:
- NEIGHBORS case
  - no list chunks, instead data by single seed position (from click), response has NEIGHBORS header

Question:
- request multiple LODs in one request?
- tracingstore vs datastore? should both offer this route? frontend needs to distinguish? (meshfiles are only on datastore. for editable mappings, datastore requests segment list from tracingstore)

Alternative thought
- new route for loading the whole mesh (backend puzzles it together)
- or just use the old routes in libs

### PR DESCRIPTION
**Pull request by fm3**
- Adds possibility to download meshes as STL (both for meshfile, ad-hoc, both for annotations and view modes, both for mappings and oversegmentations): `POST <datastore_uri>/data/datasets/:organizationName/:datasetName/layers/:dataLayerName/meshes/fullMesh.stl` and `POST <tracingstore_uri>/tracings/volume/:tracingId/fullMesh.stl` with json body:
```
     meshFileName: Option[String], // None means ad-hoc meshing
    lod: Option[Int],
    segmentId: Long, // if mappingName is set, this is an agglomerate id
    mappingName: Option[String],
    mappingType: Option[String], // json, agglomerate, editableMapping
    editableMappingTracingId: Option[String],
    mag: Option[Vec3Int], // required for ad-hoc meshing
    seedPosition: Option[Vec3Int], // required for ad-hoc meshing
    additionalCoordinates: Option[Seq[AdditionalCoordinate]]
```
- Introduces subproject webknossosNative which makes use of JNI to provide native functions (currently only one: a dracoToStlConverter)
- Some renamings to unify things: webKnossos → webknossos in some more places in the backend code and adHocMeshing → adHocMesh
- new dependency draco
- removes the subsamplingStrides feature for ad-hoc meshes. This was already unused. Compare [internal discussion](https://scm.slack.com/archives/C5AKLAV0B/p1708072916535039?thread_ts=1708008766.862249&cid=C5AKLAV0B).

### URL of deployed dev instance (used for testing):
- https://loadfullmesh.webknossos.xyz

### Steps to test:
- Create volume annotation, brush some, test via `curl 'http://localhost:9000/tracings/volume/71a2eaeb-3413-4c88-a99c-45547e6d9de5/fullMesh.stl?token=secretSampleUserToken' -X POST -H 'Accept-Encoding: gzip, deflate, br' -H 'content-type: application/json' --data-raw '{"mag": [1,1,1], "seedPosition": [99, 102, 81], "segmentId": 1}'  > out.stl` (insert your tracing id, seed position)
- Test the datastore route as well
- Test other combinations (static mappings, editable mappings, brushed on mapping, meshfile is for oversegmentation, meshfile is for target mapping, other mags, other segment ids)
- Inspect downloaded stl files for example with `meshlab` on ubuntu

### TODOs:
- [x] Implementation for Meshfile
  - [x] read mesh chunk infos
  - [x] decode draco
  - [x] serialize as headless stl
  - [x] add single stl header
  - [x] c++ error handling with try/catch, create java exception?
  - [x] get a review by someone who is more familiar with c++
  - [x] transform geometry with position
    - [x] (+matrix?)
  - [x] mapping support (also editable)
    - [x] do not map ids if meshfile is already computed for this mapping
- [x] Implementation for ad-hoc meshes
  - [x] Encode ad-hoc mesh float buffer as stl
  - [x] continue with neighbors
  - [x] glue together the chunks better (overlap/merge vertices?)
  - [x] mapping support (not editable, that goes to the tracingstore)
- [x] Should also be available in tracingstore
  - [x] Ad-Hoc
    - [x] Should use segment index for meshing if available
    - [x] how to get dataset voxel size?
      - [x] Cache timeout?
    - [x] Mapping support
      - [x] Test after changing an existing mapped segment by brushing
      - [x] Editable Mappings
  - [x] With File: forward to datastore
    - [x] with mapping support
    - [x] Editable Mappings
- [x] Measure perf
- [x] test JNI stuff on mac
- [x] Cleanup
- [x] document JNI development process

### Follow-up
- Parameter range listing route
- Python client
- Pass all infos to Render Animation Job

### Issues:
- fixes #7529

------
- [x] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [x] Updated [migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [x] Updated [documentation](../blob/master/docs) if applicable
- [x] Removed dev-only changes like prints and application.conf edits
- [x] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)
- [x] Needs datastore update after deployment

### CODE REVIEW

**Review by normanrz — COMMENTED** [review]
Code looks great! I'll do some testing next.

**Comment by normanrz** [comment]
Not sure if it is just the rendering in my Mac or if the normals in the STL are wrong
Screenshot 2024-03-08 at 12 00 38

File: build.sbt [hunk 1]
```diff
@@ -57,11 +57,17 @@ lazy val util = (project in file("util")).settings(
   dependencyOverrides ++= Dependencies.dependencyOverrides
 )
 
+lazy val webknossosNative = (project in file("webknossos-native")) // native code and build script
```
[the remark this answers is not in the document]
Comment (fm3) [reply build.sbt:60]: No strong opinion, I can change it if you want. Why would you prefer it?

[the remark this answers is not in the document]
Comment (normanrz) [reply build.sbt:60]: I thought JNI is a bit more specific. Native can mean a lot of things. No strong opinion, either.

**Review by normanrz — APPROVED** [review]

**Pull request merged**